src/hotspot/cpu/s390/macroAssembler_s390.cpp

```diff
@@ -5803,3 +5803,48 @@ void MacroAssembler::lightweight_unlock(Register obj, Register hdr, Register tmp
   z_alsi(in_bytes(JavaThread::lock_stack_top_offset()), Z_thread, -oopSize);  // pop object
   z_cr(tmp, tmp); // set CC to EQ
 }
+
+void MacroAssembler::pop_count_int(Register r_dst, Register r_src, Register r_tmp) {
+  BLOCK_COMMENT("pop_count_int {");
+
+  if (VM_Version::has_MiscInstrExt3()) {
+    z_llgfr(r_dst, r_src);
+    z_popcnt(r_dst, r_dst, 8);
+  } else {
+
+    assert(r_tmp != noreg, "temp register required for popcnt, for machines < z15");
+    assert_different_registers(r_dst, r_tmp); // if r_src is same as r_tmp, it should be fine
```

Comment:
Please move these assertions out of the if block. We want to ensure this for testing purposes at all times, so that even if we test on Z15 we expect it to run on others.

Reply:
@theRealAph,
I understand your argument. Nonetheless, I'm not happy with this request. The match rules in s390.ad have been designed such that, in case the more potent instruction is available, the register allocator does not need to provide a temp register.
Regards, Lutz

Reply:
I agree with Lutz. I have added a separate match rule which will be applicable only for machines `Z15 onwards`. Which will not pass `r_tmp` and it will be considered `noreg` will not satisfy first assert requirement.

> so that even if we test on Z15 we expect it to run on others.

`else` block will never run on `Z15 or Z16` machines.

Reply:
I see. I wasn't so much thinking about the match rules, but anywhere else that needs to use popcount. While usages from the ad file are undoubtedly safe, anywhere else must provide a register to use or have some sort of conditionals at the point pop_count_int is used.

Reply:
>anywhere else must provide a register to use or have some sort of conditionals at the point pop_count_int is used.

So for machines `<=Z14` we will never go in the `if` block; And the moment we enter in `else` block there is check, which confirms `r_tmp` shouldn't be `noreg` which will force author to pass a temporary register. Isn't it ?

Reply:
> > anywhere else must provide a register to use or have some sort of conditionals at the point pop_count_int is used.
> 
> So for machines `<=Z14` we will never go in the `if` block; And the moment we enter in `else` block there is check, which confirms `r_tmp` shouldn't be `noreg` which will force author to pass a temporary register. Isn't it ?

The point is to avoid the failure where testing is done on Z15 and up, but some customers have earlier machines. A maintenance programmer uses popcount and all the testing on Z15 passes. The first person to see the failure is the customer. We're setting a trap for maintainers.

This is not a complaint about the ad file use case. It's about anywhere else that popcount is used in hand-written assembly code. I'm telling you this from painful experience on AArch64.

Reply:
There's no need to hope, make `z_popcnt` protected.

Reply:
done; please have a look at new commits;